File: src/main/java/net/atos/client/zgw/zrc/model/RolNietNatuurlijkPersoon.java
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */

package net.atos.client.zgw.zrc.model;

import java.net.URI;
import java.util.Objects;
import java.util.UUID;

import org.apache.commons.lang3.StringUtils;

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum;
import nl.info.client.zgw.zrc.model.generated.NietNatuurlijkPersoonIdentificatie;
import nl.info.client.zgw.ztc.model.generated.RolType;

/**
 * Manually copied from {@link nl.info.client.zgw.zrc.model.generated.RolNietNatuurlijkPersoon} and modified to allow for
 * polymorphism using a generic base {@link Rol} class.
 * Ideally we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
public class RolNietNatuurlijkPersoon extends Rol<NietNatuurlijkPersoonIdentificatie> {

    public RolNietNatuurlijkPersoon() {
    }

    public RolNietNatuurlijkPersoon(
            final URI zaak,
            final RolType roltype,
            final String roltoelichting,
            final NietNatuurlijkPersoonIdentificatie betrokkeneIdentificatie
    ) {
        super(zaak, roltype, BetrokkeneTypeEnum.NIET_NATUURLIJK_PERSOON, betrokkeneIdentificatie, roltoelichting);
    }

    /**
     * For testing purposes only where we need a UUID.
     */
    public RolNietNatuurlijkPersoon(
            final UUID uuid,
            final RolType roltype,
            final String roltoelichting,
            final NietNatuurlijkPersoonIdentificatie betrokkeneIdentificatie
    ) {
        super(uuid, roltype, BetrokkeneTypeEnum.NIET_NATUURLIJK_PERSOON, betrokkeneIdentificatie, roltoelichting);
    }

    @Override
    protected boolean equalBetrokkeneIdentificatie(final NietNatuurlijkPersoonIdentificatie identificatie) {
        final NietNatuurlijkPersoonIdentificatie betrokkeneIdentificatie = getBetrokkeneIdentificatie();
        if (betrokkeneIdentificatie == identificatie) {
            return true;
        }
        if (identificatie == null || betrokkeneIdentificatie == null) {
            return false;
        }
        if (betrokkeneIdentificatie.getAnnIdentificatie() != null || identificatie.getAnnIdentificatie() != null) {
            return Objects.equals(betrokkeneIdentificatie.getAnnIdentificatie(), identificatie.getAnnIdentificatie());
        }
        if (betrokkeneIdentificatie.getInnNnpId() != null || identificatie.getInnNnpId() != null) {
            return Objects.equals(betrokkeneIdentificatie.getInnNnpId(), identificatie.getInnNnpId());
        }
        if (betrokkeneIdentificatie.getVestigingsNummer() != null || identificatie.getVestigingsNummer() != null) {
            return Objects.equals(betrokkeneIdentificatie.getVestigingsNummer(), identificatie.getVestigingsNummer());
        }
        return true;
    }

    @Override
    public String getNaam() {
        if (getBetrokkeneIdentificatie() == null) {
            return null;
        }
        return StringUtils.isNotEmpty(getBetrokkeneIdentificatie().getStatutaireNaam()) ?
                getBetrokkeneIdentificatie().getStatutaireNaam() : getIdentificatienummer();
    }

    @Override
    public String getIdentificatienummer() {
        NietNatuurlijkPersoonIdentificatie identificatie = getBetrokkeneIdentificatie();
        if (identificatie == null) {
            return null;
        }
        if (StringUtils.isNotEmpty(identificatie.getAnnIdentificatie())) {
            return identificatie.getAnnIdentificatie();
        }
        if (StringUtils.isNotEmpty(identificatie.getInnNnpId())) {
            return identificatie.getInnNnpId();
        }
        return identificatie.getVestigingsNummer();
    }

    @Override
    protected int hashCodeBetrokkeneIdentificatie() {
        NietNatuurlijkPersoonIdentificatie identificatie = getBetrokkeneIdentificatie();
        if (identificatie == null) {
            return 0;
        }
        if (identificatie.getAnnIdentificatie() != null) {
            return Objects.hash(getBetrokkeneIdentificatie().getAnnIdentificatie());
        }
        if (identificatie.getInnNnpId() != null) {
            return Objects.hash(getBetrokkeneIdentificatie().getInnNnpId());
        }
        if (identificatie.getVestigingsNummer() != null) {
            return Objects.hash(getBetrokkeneIdentificatie().getVestigingsNummer());
        }
        return 0;
    }
}

File: src/main/kotlin/nl/info/client/zgw/zrc/model/RolNietNatuurlijkPersoon.kt
/*
 * SPDX-FileCopyrightText: 2021 Atos, 2023, 2026 INFO.nl
 * SPDX-License-Identifier: EUPL-1.2+
 */
package nl.info.client.zgw.zrc.model

import nl.info.client.zgw.zrc.model.generated.BetrokkeneTypeEnum
import nl.info.client.zgw.zrc.model.generated.NietNatuurlijkPersoonIdentificatie
import nl.info.client.zgw.ztc.model.generated.RolType
import org.apache.commons.lang3.StringUtils
import java.net.URI
import java.util.Objects
import java.util.UUID

/**
 * Manually copied from [nl.info.client.zgw.zrc.model.generated.RolNietNatuurlijkPersoon] and modified to allow for
 * polymorphism using a generic base [Rol] class.
 * Ideally, we would use the generated class, but currently we cannot get the OpenAPI Generator framework to generate
 * polymorphic relationships correctly.
 */
class RolNietNatuurlijkPersoon : Rol<NietNatuurlijkPersoonIdentificatie> {
    constructor() : super()

    constructor(
        zaak: URI,
        roltype: RolType,
        roltoelichting: String,
        betrokkeneIdentificatie: NietNatuurlijkPersoonIdentificatie?
    ) : super(zaak, roltype, BetrokkeneTypeEnum.NIET_NATUURLIJK_PERSOON, betrokkeneIdentificatie, roltoelichting)

    /**
     * For testing purposes only where we need a UUID.
     */
    constructor(
        uuid: UUID,
        roltype: RolType,
        roltoelichting: String,
        betrokkeneIdentificatie: NietNatuurlijkPersoonIdentificatie?
    ) : super(uuid, roltype, BetrokkeneTypeEnum.NIET_NATUURLIJK_PERSOON, betrokkeneIdentificatie, roltoelichting)

    @Suppress("ReturnCount")
    override fun equalBetrokkeneIdentificatie(identificatie: NietNatuurlijkPersoonIdentificatie?): Boolean {
        val ownIdentificatie = betrokkeneIdentificatie
        if (ownIdentificatie === identificatie) {
            return true
        }
        if (identificatie == null || ownIdentificatie == null) {
            return false
        }
        if (ownIdentificatie.innNnpId != null || identificatie.innNnpId != null) {
            return Objects.equals(ownIdentificatie.innNnpId, identificatie.innNnpId)
        }
        if (ownIdentificatie.kvkNummer != null || identificatie.kvkNummer != null) {
            return Objects.equals(ownIdentificatie.kvkNummer, identificatie.kvkNummer)
        }
        if (ownIdentificatie.vestigingsNummer != null || identificatie.vestigingsNummer != null) {
            return Objects.equals(ownIdentificatie.vestigingsNummer, identificatie.vestigingsNummer)
        }
        if (ownIdentificatie.annIdentificatie != null || identificatie.annIdentificatie != null) {
            return Objects.equals(ownIdentificatie.annIdentificatie, identificatie.annIdentificatie)
        }
        return true
    }

    override val naam: String?
        get() {
            val identificatie = betrokkeneIdentificatie ?: return null
            return if (StringUtils.isNotEmpty(identificatie.statutaireNaam)) identificatie.statutaireNaam else identificatienummer
        }

    override val identificatienummer: String?
        get() {
            val identificatie = betrokkeneIdentificatie ?: return null
            return when {
                // new 'RSIN-type' initiators only have a KVK number (but no vestigingsnummer)
                StringUtils.isNotBlank(identificatie.kvkNummer) && StringUtils.isBlank(identificatie.vestigingsNummer) ->
                    identificatie.kvkNummer
                // we also support 'legacy' RSIN-type initiators with only an RSIN (no KVK number)
                StringUtils.isNotEmpty(identificatie.innNnpId) -> identificatie.innNnpId
                // lastly we support the 'vestiging-type' initiators that have both a KVK number and a vestigingsnummer
                // note that the KVK number is not part of this class but returned differently
                else -> identificatie.vestigingsNummer
            }
        }

    override fun hashCodeBetrokkeneIdentificatie(): Int {
        val identificatie = betrokkeneIdentificatie ?: return 0
        return when {
            identificatie.innNnpId != null -> Objects.hash(identificatie.innNnpId)
            identificatie.kvkNummer != null -> Objects.hash(identificatie.kvkNummer)
            identificatie.vestigingsNummer != null -> Objects.hash(identificatie.vestigingsNummer)
            identificatie.annIdentificatie != null -> Objects.hash(identificatie.annIdentificatie)
            else -> 0
        }
    }
}
